### DISCUSSION TITLE
Version 13.4.0+ breaks tests running on Azure DevOps Agent on CentOS

### DISCUSSION THREAD

**Discussion by KTL52** [cypress-io/cypress#28399] [Questions and Help]
Hey,

I have been having an issue recently where our Cypress tests that runs on an Azure DevOps agent deployed on a CentOS machine are unable to run. A snippet of the console output is attached here [output.txt](https://github.com/cypress-io/cypress/files/13456811/output.txt).

Following [https://docs.cypress.io/guides/continuous-integration/introduction#Dependencies](url) i do have the required dependencies installed. 

There are no issues when i run the exact same tests on Cypress version 13.3.3.
When i run the tests locally on Windows (on all Cypress versions) there are also no issues.

Update:
Clearing the cache (removing the /kpg/.cache/Cypress/13.4.0/) seems to resolve the issue. Does anyone have an idea on how this can be prevented again in the future?

**Comment by portyu9** [comment]
Hi @KTL52 

Since deleting:

```text
/kpg/.cache/Cypress/13.4.0/
```

fixed the problem, the strongest evidence points to that cached Cypress binary being incomplete/corrupted rather than your test suite or CentOS dependencies.

To make CI more robust, I would:

1. Install dependencies cleanly:
   ```bash
   npm ci
   ```

2. Cache the Cypress binary directory:
   ```text
   ~/.cache/Cypress
   ```

3. **Do not cache `node_modules` directly.** Cache the npm/yarn/pnpm package cache instead.

4. Key the Cypress cache to the lockfile/Cypress version so a new Cypress version cannot accidentally reuse an incompatible cache.

5. Add diagnostics before the test:
   ```bash
   npx cypress cache path
   npx cypress cache list
   npx cypress verify
   ```

6. If a restored cache is missing the expected binary, explicitly repair it:
   ```bash
   npx cypress install
   ```

If you want to aggressively recover from a bad cache:

```bash
npx cypress cache clear
npx cypress install
npx cypress verify
```

I would not clear the cache on every build because that removes the performance benefit. Instead, make the cache key version/lockfile-specific and only invalidate it when verification fails or dependencies change.

Please mark this answer as accepted if it helped, thank you.
